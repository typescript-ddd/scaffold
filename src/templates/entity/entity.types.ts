import {
  TemplateBooleanOption,
  TemplatePropertiesOption,
  TemplatePropertyValue,
  TemplateStringOption,
} from "../shared/template";

export type EntityTemplateValues = {
  entityName: string;
  properties: TemplatePropertyValue[];
  trackable: boolean;
};

export type EntityTemplateOptions = {
  entityName: TemplateStringOption;
  properties: TemplatePropertiesOption;
  trackable: TemplateBooleanOption;
};
