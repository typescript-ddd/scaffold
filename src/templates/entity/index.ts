export * from "./entity.generator";
export * from "./entity.template";
export * from "./entity.types";
