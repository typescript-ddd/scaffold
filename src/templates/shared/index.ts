export * from "./context-builder";
export * from "./generate-context";
export * from "./template";