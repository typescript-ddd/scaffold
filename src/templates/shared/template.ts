import { GenerateContext } from "./generate-context";

export type TemplateEnum = Record<string, string>;
export type TemplatePropertyValueType =
  | "string"
  | "number"
  | "boolean"
  | "date";

export type TemplateEnumOption = {
  name: string;
  label: string;
  description: string;
  enum: TemplateEnum;
  optionType: "enum";
};

export type TemplateStringOption = {
  name: string;
  label: string;
  description: string;
  optionType: "string";
};

export type TemplateTokenStringOption = {
  name: string;
  label: string;
  description: string;
  optionType: "tokenString";
};

export type TemplateBooleanOption = {
  name: string;
  label: string;
  description: string;
  optionType: "boolean";
};

export type TemplateNumberOption = {
  name: string;
  label: string;
  description: string;
  optionType: "number";
};

export type TemplateDateOption = {
  name: string;
  label: string;
  description: string;
  optionType: "date";
};

export type TemplatePropertyOption = {
  name: string;
  valueType: TemplatePropertyValueType;
  nullable?: boolean;
  prop?: string;
};

export type TemplatePropertyValue = {
  name: string;
  valueType: string;
  nullable?: boolean;
  prop?: string;
};

export type TemplatePropertiesOption = {
  name: string;
  label: string;
  description: string;
  optionType: "properties";
};

export type TemplateOption =
  | TemplateEnumOption
  | TemplateStringOption
  | TemplateBooleanOption
  | TemplateNumberOption
  | TemplateDateOption
  | TemplatePropertiesOption
  | TemplateTokenStringOption;

export type Maybe<T> = T | null | undefined;

export interface Template<
  TValues extends { [K in keyof TOptions]: any},
  TOptions extends { [key: string]: TemplateOption }
> {
  name: string;
  description: string;
  options: TOptions;
  defaultValues: TValues;
  generate(values: TValues, context: GenerateContext): string;
}