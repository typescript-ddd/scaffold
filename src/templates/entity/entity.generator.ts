import { Project, Scope, StructureKind } from "ts-morph";
import { strings } from "../../utils";
import { GenerateContext } from "../shared";
import { EntityTemplateValues } from "./entity.types";

export const generateEntity = (
  values: EntityTemplateValues,
  context: GenerateContext
) => {
  const { entityName, properties = [], trackable = false } = values;
  const name = strings.capitalize(entityName);
  const idName = `${name}Id`;
  const fileName = strings.kebab(entityName);
  const baseName = trackable ? "TrackableEntity" : "Entity";

  const project = new Project();
  const sourceFile = project.createSourceFile(
    `${fileName}.ts`,
    `
export interface Create${name}Props extends EntityCreateProps {}
export interface Update${name}Props extends EntityUpdateProps<${idName}> {}
export class ${name} extends ${baseName}<${idName}> {}    
`
  );

  context.addImportDeclaration({
    moduleSpecifier: `@typescript-ddd/core`,
    namedImports: [
      baseName,
      "EntityCreateProps",
      "EntityUpdateProps",
      trackable ? "UtcDate" : "",
    ],
  });

  context.addImportDeclaration({
    moduleSpecifier: context.resolveDir("domain"),
    namedImports: [`${name}Id`],
  });

  context.writeSourceFileImports(sourceFile);

  const createInterface = sourceFile.getInterface(`Create${name}Props`);
  createInterface.addProperties(
    properties.map((property) => ({
      name: property.name,
      type: property.valueType,
      hasQuestionToken: property.nullable,
    }))
  );
  createInterface.addJsDoc({
    description: `Represents the properties required to create a ${strings.lower(
      name
    )}.`,
    tags: [{ tagName: "extends", text: "{EntityCreateProps}" }],
  });
  if (trackable) {
    createInterface.addProperties([
      { name: "createdAtUtc", type: "UtcDate", hasQuestionToken: true },
      { name: "updatedAtUtc", type: "UtcDate", hasQuestionToken: true },
    ]);
  }
  const updateInterface = sourceFile.getInterface(`Update${name}Props`);
  updateInterface.addProperties(
    properties.map((property) => ({
      name: property.name,
      type: property.valueType,
      hasQuestionToken: property.nullable,
    }))
  );
  updateInterface.addJsDoc({
    description: `Represents the properties required to update a ${strings.lower(
      name
    )}.`,
    tags: [{ tagName: "extends", text: "{EntityUpdateProps}" }],
  });
  const classDeclaration = sourceFile.getClass(name);
  classDeclaration.addJsDoc({
    description: `Represents a ${strings.lower(name)} entity.`,
    tags: [{ tagName: "extends", text: baseName }],
  });
  const constructorDeclaration = classDeclaration.addConstructor({
    scope: Scope.Private,
    parameters: trackable
      ? [
          { name: "id", type: `${name}Id`, hasQuestionToken: true },
          ...properties.map((property) => ({
            name: property.name,
            type: property.valueType,
            scope: Scope.Private,
            hasQuestionToken: property.nullable,
          })),
          { name: "createdAtUtc", type: "UtcDate", hasQuestionToken: true },
          { name: "updatedAtUtc", type: "UtcDate", hasQuestionToken: true },
        ]
      : [
          { name: "id", type: `${name}Id`, hasQuestionToken: true },
          ...properties.map((property) => ({
            name: property.name,
            type: property.valueType,
            scope: Scope.Private,
            hasQuestionToken: property.nullable,
          })),
        ],
  });
  constructorDeclaration.setBodyText((writer) => {
    if (trackable) {
      writer.writeLine(
        `super(id ?? ${idName}.create(), createdAtUtc, updatedAtUtc);`
      );
    } else {
      writer.writeLine(`super(id ?? ${idName}.create());`);
    }
  });
  constructorDeclaration.addJsDoc({
    description: `Initializes a new instance of ${strings.describe(name)}.`,
    tags: [
      {
        tagName: "param",
        text: `{${name}Id} id - An optional identifier of the ${strings.lower(
          name
        )}.`,
      },
      ...properties.map((property) => ({
        tagName: "param",
        text: `{${property.valueType}${property.nullable ? "?" : ""}} ${
          property.name
        } - The ${property.name} of the ${strings.lower(name)}.`,
      })),
      ...(trackable
        ? [
            {
              tagName: "param",
              text: "{UtcDate} createdAtUtc - An optional creation date.",
            },
            {
              tagName: "param",
              text: "{UtcDate} updatedAtUtc - An optional last update date.",
            },
          ]
        : []),
    ],
  });

  properties.forEach((property) => {
    classDeclaration
      .addMethod({
        name: `get${strings.capitalize(property.name)}`,
        returnType: property.valueType,
        hasQuestionToken: property.nullable,
      })
      .setBodyText(`return this.${property.name};`)
      .addJsDoc({
        description: `Gets the ${property.name} of the ${strings.lower(name)}.`,
        tags: [
          {
            tagName: "returns",
            text: `{${property.valueType}${
              property.nullable ? "?" : ""
            }} - The ${property.name} of the ${strings.lower(name)}.`,
          },
        ],
      });
  });

  classDeclaration
    .addMethod({
      name: "update",
      returnType: "void",
      parameters: [{ name: "props", type: `Update${name}Props` }],
      docs: [
        {
          description: `Update the ${strings.natural(name)}.`,
          kind: StructureKind.JSDoc,
          tags: [
            {
              tagName: "param",
              text: `{Update${name}Props} props - The properties to update.`,
            },
            {
              tagName: "returns",
              text: "void",
            },
          ],
        },
      ],
    })
    .setBodyText((writer) => {
      writer.writeLine(`// TODO: validate props`);
      properties.forEach((property) => {
        writer.writeLine(`this.${property.name} = props.${property.name};`);
      });
      if (trackable) {
        writer.writeLine(`this.updatedAtUtc = UtcDate.now();`);
      }
      writer.writeLine(
        `this.applyChange(new ${name}UpdatedEvent(this.id, this));`
      );
    });

  let output = sourceFile.getFullText();

  return output;
};
