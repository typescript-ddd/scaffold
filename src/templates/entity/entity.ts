import { Project, Scope, StructureKind } from "ts-morph";
import { strings } from "../../utils";
import { GenerateContext } from "../shared";

export type GenerateEntityOptions = {
  entityName: string;
  properties: { name: string; type: string; nullable?: boolean }[];
  trackable?: boolean;
  importDeclarations?: { moduleSpecifier: string; namedImports: string[] }[];
};

export const generateEntity = (
  options: GenerateEntityOptions,
  context: GenerateContext
) => {
  const {
    entityName,
    properties = [],
    trackable = false,
    importDeclarations = [],
  } = options;
  const name = strings.capitalize(entityName);
  const idName = `${name}Id`;
  const fileName = strings.kebab(entityName);
  const baseName = trackable ? "TrackableEntity" : "Entity";

  const project = new Project();
  const sourceFile = project.createSourceFile(
    `${fileName}.ts`,
    `
export interface Create${name}Props extends EntityCreateProps {}
export interface Update${name}Props extends EntityUpdateProps<${idName}> {}
export class ${name} extends ${baseName}<${idName}> {}    
`
  );

  sourceFile.addImportDeclaration({
    moduleSpecifier: `@typescript-ddd/core`,
    namedImports: [
      baseName,
      "EntityCreateProps",
      "EntityUpdateProps",
      trackable ? "UtcDate" : "",
      ...importDeclarations
        .filter((x) => x.moduleSpecifier === "@typescript-ddd/core")
        .map((x) => x.namedImports)
        .flat(),
    ],
  });
  const modelsImportDir = context.resolveDir("domain") + `/models`;
  sourceFile.addImportDeclaration({
    moduleSpecifier: modelsImportDir,
    namedImports: [
      `${name}Id`,
      ...importDeclarations
        .filter((x) => x.moduleSpecifier === modelsImportDir)
        .map((x) => x.namedImports)
        .flat(),
    ],
  });

  for (const importDeclaration of importDeclarations.filter(
    (x) =>
      !["@typescript-ddd/core", modelsImportDir].includes(x.moduleSpecifier)
  )) {
    sourceFile.addImportDeclaration(importDeclaration);
  }

  const createInterface = sourceFile.getInterface(`Create${name}Props`);
  createInterface.addProperties(
    properties.map((property) => ({
      name: property.name,
      type: property.type,
      hasQuestionToken: property.nullable,
    }))
  );
  createInterface.addJsDoc({
    description: `Represents the properties required to create a ${strings.lower(
      name
    )}.`,
    tags: [{ tagName: "extends", text: "{EntityCreateProps}" }],
  });
  if (trackable) {
    createInterface.addProperties([
      { name: "createdAtUtc", type: "UtcDate", hasQuestionToken: true },
      { name: "updatedAtUtc", type: "UtcDate", hasQuestionToken: true },
    ]);
  }
  const updateInterface = sourceFile.getInterface(`Update${name}Props`);
  updateInterface.addProperties(
    properties.map((property) => ({
      name: property.name,
      type: property.type,
      hasQuestionToken: property.nullable,
    }))
  );
  updateInterface.addJsDoc({
    description: `Represents the properties required to update a ${strings.lower(
      name
    )}.`,
    tags: [{ tagName: "extends", text: "{EntityUpdateProps}" }],
  });
  const classDeclaration = sourceFile.getClass(name);
  classDeclaration.addJsDoc({
    description: `Represents a ${strings.lower(name)} entity.`,
    tags: [{ tagName: "extends", text: baseName }],
  });
  const constructorDeclaration = classDeclaration.addConstructor({
    scope: Scope.Private,
    parameters: trackable
      ? [
          { name: "id", type: `${name}Id`, hasQuestionToken: true },
          ...properties.map((property) => ({
            name: property.name,
            type: property.type,
            scope: Scope.Private,
            hasQuestionToken: property.nullable,
          })),
          { name: "createdAtUtc", type: "UtcDate", hasQuestionToken: true },
          { name: "updatedAtUtc", type: "UtcDate", hasQuestionToken: true },
        ]
      : [
          { name: "id", type: `${name}Id`, hasQuestionToken: true },
          ...properties.map((property) => ({
            name: property.name,
            type: property.type,
            scope: Scope.Private,
            hasQuestionToken: property.nullable,
          })),
        ],
  });
  constructorDeclaration.setBodyText((writer) => {
    if (trackable) {
      writer.writeLine(
        `super(id ?? ${idName}.create(), createdAtUtc, updatedAtUtc);`
      );
    } else {
      writer.writeLine(`super(id ?? ${idName}.create());`);
    }
  });
  constructorDeclaration.addJsDoc({
    description: `Initializes a new instance of ${strings.describe(name)}.`,
    tags: [
      {
        tagName: "param",
        text: `{${name}Id} id - An optional identifier of the ${strings.lower(
          name
        )}.`,
      },
      ...properties.map((property) => ({
        tagName: "param",
        text: `{${property.type}${property.nullable ? "?" : ""}} ${
          property.name
        } - The ${property.name} of the ${strings.lower(name)}.`,
      })),
      ...(trackable
        ? [
            {
              tagName: "param",
              text: "{UtcDate} createdAtUtc - An optional creation date.",
            },
            {
              tagName: "param",
              text: "{UtcDate} updatedAtUtc - An optional last update date.",
            },
          ]
        : []),
    ],
  });

  properties.forEach((property) => {
    classDeclaration
      .addMethod({
        name: `get${strings.capitalize(property.name)}`,
        returnType: property.type,
        hasQuestionToken: property.nullable,
      })
      .setBodyText(`return this.${property.name};`)
      .addJsDoc({
        description: `Gets the ${property.name} of the ${strings.lower(name)}.`,
        tags: [
          {
            tagName: "returns",
            text: `{${property.type}${property.nullable ? "?" : ""}} - The ${
              property.name
            } of the ${strings.lower(name)}.`,
          },
        ],
      });
  });

  classDeclaration
    .addMethod({
      name: "update",
      returnType: "void",
      parameters: [{ name: "props", type: `Update${name}Props` }],
      docs: [
        {
          description: `Update the ${strings.natural(name)}.`,
          kind: StructureKind.JSDoc,
          tags: [
            {
              tagName: "param",
              text: `{Update${name}Props} props - The properties to update.`,
            },
            {
              tagName: "returns",
              text: "void",
            },
          ],
        },
      ],
    })
    .setBodyText((writer) => {
      writer.writeLine(`// TODO: validate props`);
      properties.forEach((property) => {
        writer.writeLine(`this.${property.name} = props.${property.name};`);
      });
      if (trackable) {
        writer.writeLine(`this.updatedAtUtc = UtcDate.now();`);
      }
      writer.writeLine(
        `this.applyChange(new ${name}UpdatedEvent(this.id, this));`
      );
    });

  let output = sourceFile.getFullText();

  return output;
};
