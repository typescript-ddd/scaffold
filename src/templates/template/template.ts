import { Project, Scope } from "ts-morph";
import { strings } from "../../utils";
import { GenerateContext } from "../shared";

export type GenerateTemplateOptions = {
  name: string;
};

export const generateTemplate = (
  options: GenerateTemplateOptions,
  context: GenerateContext
) => {
  const { name } = options;
  const className = strings.capitalize(name);

  const project = new Project();
  const sourceFile = project.createSourceFile(
    `${context.fileName}.ts`,
    `
export class ${className} {}    
`
  );

  context.addImportDeclaration({
    moduleSpecifier: `@typescript-ddd/core`,
    namedImports: ["Entity"],
  });

  context.writeSourceFileImports(sourceFile);

  const classDeclaration = sourceFile.getClass(className);
  classDeclaration.addJsDoc({
    description: `${strings.describe(name)}.`,
  });

  const constructorDeclaration = classDeclaration.addConstructor({
    scope: Scope.Private,
    parameters: [{ name: "value", type: "string" }],
  });
  constructorDeclaration.setBodyText(`super(value);`);
  constructorDeclaration.addJsDoc({
    description: `Initializes a new instance of ${strings.describe(
      name
    )} identifier.`,
    tags: [{ tagName: "param", text: "value The value of the identifier." }],
  });

  let output = sourceFile.getFullText();

  return output;
};
