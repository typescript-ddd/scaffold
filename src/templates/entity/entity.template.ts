import { Chunk, GenerateContext } from "../shared";
import { Template } from "../shared/template";
import { generateEntity } from "./entity.generator";
import { EntityTemplateOptions, EntityTemplateValues } from "./entity.types";

export class EntityTemplate
  implements Template<EntityTemplateValues, EntityTemplateOptions>
{
  readonly name = "Command";
  readonly description = "Generates a command";
  readonly options: EntityTemplateOptions = {
    entityName: {
      name: "entityName",
      label: "Entity Name",
      description: "The name of the entity",
      optionType: "string",
    },
    properties: {
      name: "properties",
      label: "Properties",
      description: "The properties of the command",
      optionType: "properties",
    },
    trackable: {
      name: "trackable",
      label: "Trackable",
      description: "Whether the entity is trackable",
      optionType: "boolean",
    },
  };
  readonly defaultValues = {
    entityName: "User",
    properties: [{ name: "name", valueType: "string" }],
    trackable: false,
  };
  generate(values: EntityTemplateValues, context: GenerateContext, chunkName?: string): Chunk {
    return generateEntity(values, context, chunkName);
  }
}
