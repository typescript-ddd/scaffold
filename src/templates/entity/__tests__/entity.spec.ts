import { ContextBuilder, GenerateContext } from "../../shared";
import { EntityTemplate } from "../entity.template";

describe("Entity", () => {
  let context: GenerateContext;
  let template: EntityTemplate;

  beforeEach(() => {
    context = ContextBuilder.build("@src");
    context.addImportDeclaration({
      moduleSpecifier: context.resolveDir("domain"),
      namedImports: ["OrderId"],
    });
    template = new EntityTemplate();
  });

  it("Should render the entity component", () => {
    const output = template.generate(
      {
        entityName: "Invoice",
        properties: [{ name: "orderId", valueType: "OrderId" }],
        trackable: false,
      },
      context
    );

    expect(output).toMatchSnapshot();
  });

  it("Should render the trackable entity component", () => {
    const context = ContextBuilder.build("@src");
    const output = template.generate(
      {
        entityName: "Invoice",
        properties: [{ name: "orderId", valueType: "OrderId" }],
        trackable: true,
      },
      context
    );

    expect(output).toMatchSnapshot();
  });
});
