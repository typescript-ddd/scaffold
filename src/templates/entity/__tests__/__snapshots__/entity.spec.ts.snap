// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Entity Should render the entity component 1`] = `
"import { Entity, EntityCreateProps, EntityUpdateProps  } from "@typescript-ddd/core";
import { OrderId, InvoiceId } from "@src/domain";

/**
 * Represents the properties required to create a invoice.
 * @extends {EntityCreateProps}
 */
export interface CreateInvoiceProps extends EntityCreateProps {
    orderId: OrderId;
}
/**
 * Represents the properties required to update a invoice.
 * @extends {EntityUpdateProps}
 */
export interface UpdateInvoiceProps extends EntityUpdateProps<InvoiceId> {
    orderId: OrderId;
}
/**
 * Represents a invoice entity.
 * @extends Entity
 */
export class Invoice extends Entity<InvoiceId> {
    /**
     * Initializes a new instance of a invoice.
     * @param {InvoiceId} id - An optional identifier of the invoice.
     * @param {OrderId} orderId - The orderId of the invoice.
     */
    private constructor(id?: InvoiceId, private orderId: OrderId) {
        super(id ?? InvoiceId.create());
    }

    /**
     * Gets the orderId of the invoice.
     * @returns {OrderId} - The orderId of the invoice.
     */
    getOrderId(): OrderId {
        return this.orderId;
    }

    /**
     * Update the invoice.
     * @param {UpdateInvoiceProps} props - The properties to update.
     * @returns void
     */
    update(props: UpdateInvoiceProps): void {
        // TODO: validate props
        this.orderId = props.orderId;
        this.applyChange(new InvoiceUpdatedEvent(this.id, this));
    }
}    
"
`;

exports[`Entity Should render the trackable entity component 1`] = `
"import { TrackableEntity, EntityCreateProps, EntityUpdateProps, UtcDate } from "@typescript-ddd/core";
import { InvoiceId } from "@src/domain";

/**
 * Represents the properties required to create a invoice.
 * @extends {EntityCreateProps}
 */
export interface CreateInvoiceProps extends EntityCreateProps {
    orderId: OrderId;
    createdAtUtc?: UtcDate;
    updatedAtUtc?: UtcDate;
}
/**
 * Represents the properties required to update a invoice.
 * @extends {EntityUpdateProps}
 */
export interface UpdateInvoiceProps extends EntityUpdateProps<InvoiceId> {
    orderId: OrderId;
}
/**
 * Represents a invoice entity.
 * @extends TrackableEntity
 */
export class Invoice extends TrackableEntity<InvoiceId> {
    /**
     * Initializes a new instance of a invoice.
     * @param {InvoiceId} id - An optional identifier of the invoice.
     * @param {OrderId} orderId - The orderId of the invoice.
     * @param {UtcDate} createdAtUtc - An optional creation date.
     * @param {UtcDate} updatedAtUtc - An optional last update date.
     */
    private constructor(id?: InvoiceId, private orderId: OrderId, createdAtUtc?: UtcDate, updatedAtUtc?: UtcDate) {
        super(id ?? InvoiceId.create(), createdAtUtc, updatedAtUtc);
    }

    /**
     * Gets the orderId of the invoice.
     * @returns {OrderId} - The orderId of the invoice.
     */
    getOrderId(): OrderId {
        return this.orderId;
    }

    /**
     * Update the invoice.
     * @param {UpdateInvoiceProps} props - The properties to update.
     * @returns void
     */
    update(props: UpdateInvoiceProps): void {
        // TODO: validate props
        this.orderId = props.orderId;
        this.updatedAtUtc = UtcDate.now();
        this.applyChange(new InvoiceUpdatedEvent(this.id, this));
    }
}    
"
`;
